Do not adapt across variants with different roles

bhh1988 opened the pull request:
This fixes a bug where two audio tracks with the same language but
different roles might be switched in and out due to ABR.

Technically the DASH spec requires that the player not adapt in/out of a
single AdaptationSet, but since the "variant" concept does not track the
AdaptationSet that its component streams originate from, the closest
thing we can go by is the "role", since two streams with different roles
certainly come from different adaptation sets.

Issue #947

ismena commented:
LGTM, let me run this through the build bot.

vaage commented:
We are in the process of updating the build bot, this failure has to do with some of the changes we made and not your change.

vaage commented:
Could you rebase your change on top of master? We updated `build/test.py` and that is why the build bot is failing to build right now.

bhh1988 commented:
Ok I rebased and forced update on my branch. Do I have to create a new pull-request?

vaage commented:
@ismena The build bot is happy with this change. Can I pass this back over to you to finish up with?

bhh1988 commented:
Looks good. Can someone merge this? Then please look at https://github.com/google/shaka-player/pull/950

vaage commented:
@ismena I am not ready to accept this change for two reasons:
1. There is no test that reproducibles the original problem. I would like to see a test that shows that the ABR system will adapt between roles without this fix and show that it won't with this fix. Without that we cannot assert that this change fixes that problem.
2.  This only part of a bigger problem. In #918 I point that a problem in how ABR could change between languages. I feel that this change is only treating a symptom and not treating the root problem.

bhh1988 commented:
Thanks for taking a look @vaage 

I've pushed a new update, addressing your concerns. I wasn't aware of #918 at the time, thanks for pointing that out - I have fixed that issue as well since it was somewhat similar. I recognize that the filterTextStreams and filterVariants code look nearly identical and was hoping to merge them to address #858 as well,  but unfortunately the current logic is different enough (due to there being 2 streams, audio/video, for variants, versus 1 stream for text) that I decided to leave that much alone.

As far as testing goes, I added more unit tests that cover pretty much all the combinations of language, role, primary-state. These tests would've failed without my changes, and therefore should catch any regressions. However I did not write tests that explicitly test ABR, because my changes are technically isolated to the filter functions, which fix the ABR issue only incidentally. Testing for the ABR issue would be an integration test instead of a unit test, and I'm not familiar enough yet with the code-base to add such a thing soon. If you still feel like an integration test should be added, perhaps another issue can be filed?

Also, can someone answer my question in #947 ? I'm questioning whether it was intended to hide adaptation-set IDs from the variants?

vaage commented:
@bhh1988 I have been working on addressing #858. We have a change in review right now and it we have been going back-and-forth on it a lot - so don't worry about trying to resolve that.

Thank you for adding more tests, I will take a look at them today.

@joeyparrish could you answer the question relating to this in #947? It requires more historic knowledge than I have.

I will also talk with @joeyparrish about how we want to address the root issue.

vaage commented:
@bhh1988 I have talked with @joeyparrish and it looks like we are going to take on this issue in-house as it ties into a larger problem. We thank you for identifying this problem and I assure you that we will resolve this issue promptly.
